pkg/coscheduling/core/core_test.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	gochache "github.com/patrickmn/go-cache"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	clientsetfake "k8s.io/client-go/kubernetes/fake"
+	clicache "k8s.io/client-go/tools/cache"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+	st "k8s.io/kubernetes/pkg/scheduler/testing"
+
+	"sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
+	fakepgclientset "sigs.k8s.io/scheduler-plugins/pkg/generated/clientset/versioned/fake"
+	pgformers "sigs.k8s.io/scheduler-plugins/pkg/generated/informers/externalversions"
+	"sigs.k8s.io/scheduler-plugins/pkg/util"
+	testutil "sigs.k8s.io/scheduler-plugins/test/util"
+)
+
+func TestPreFilter(t *testing.T) {
+	ctx := context.Background()
+	cs := fakepgclientset.NewSimpleClientset()
+
+	pgInformerFactory := pgformers.NewSharedInformerFactory(cs, 0)
+	pgInformer := pgInformerFactory.Scheduling().V1alpha1().PodGroups()
+	pgInformerFactory.Start(ctx.Done())
+
+	pg := makePG("pg", 2)
+	pg1 := makePG("pg1", 2)
+	pgInformer.Informer().GetStore().Add(pg)
+	pgInformer.Informer().GetStore().Add(pg1)
+	pgLister := pgInformer.Lister()
+	denyCache := newCache()
+	denyCache.SetDefault("ns1/pg1", "ns1/pg1")
+
+	tests := []struct {
+		name            string
+		pod             *corev1.Pod
+		pods            []*corev1.Pod
+		lastDeniedPG    *gochache.Cache
+		expectedSuccess bool
+	}{
+		{
+			name: "pod does not belong to any pg",
+			pod:  st.MakePod().Name("p").UID("p").Namespace("ns1").Obj(),
+			pods: []*corev1.Pod{
+				st.MakePod().Name("pg1-1").UID("pg1-1").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+				st.MakePod().Name("pg2-1").UID("pg2-1").Namespace("ns1").Label(util.PodGroupLabel, "pg2").Obj(),
+			},
+			lastDeniedPG:    newCache(),
+			expectedSuccess: true,
+		},
+		{
+			name:            "pg was previously denied",
+			pod:             st.MakePod().Name("p1").UID("p1").Namespace("ns1").Label(util.PodGroupLabel, "pg1").Obj(),
+			lastDeniedPG:    denyCache,
```

Comment:
lastDeniedPG would make sense when a job failed, this would help avoid pods' frequently failing and go to `Filter` logic.
IMO side effect is almost none the freeze time is very small.